### PR TITLE
Lecture 02 Notes

### PR DESCRIPTION
**Pull request by cphalen**

### CODE REVIEW

File: content/lectures/lec02.md [hunk 1]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
 draft: false
 ---
 
+# Lecture Two: Docker
+
+## Portability
+
+Recall all of the issues that we've run into trying transfer code that we developed on our local machine to a remote host in the cloud. We want to be able to take our application code, and seamlessly run it on another machine. Along with Python dependencies, we needed to make sure that system resources, including Python and Poetry themselves, were also installed as expected. How can we avoid all of the dependency and versioning issues we ran into when we manually deployed to EC2?
```
Comment (pawalt) [inline content/lectures/lec02.md:13]:
```suggestion
Recall all of the issues that we've run into trying transfer code that we developed on our local machine to a remote host in the cloud. We want to be able to take our application code and seamlessly run it on another machine. Along with Python dependencies, we needed to make sure that system resources, including Python and Poetry themselves, were also installed as expected. How can we avoid all of the dependency and versioning issues we ran into when we manually deployed to EC2?
```

File: content/lectures/lec02.md [hunk 2]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+We were able to use Poetry and get all of the *reproducibility* benefits when we deployed to EC2, but just because a build is reproducible does *not* mean that it's **portable**.
+
+The intuition here is that we essentially want to _send a copy of our local computer_ to whoever is going to be running our code.
+
+## Containers
+
+Containers allow us to package minimal operating systems as "images" which includes everything our program needs to run. Once we've created the image, we can send it to anyone, and they will be able to build it on their machine to run our code. If you've used Virtual Machines in previous classes, you can think of container images as analogous to VM images, but with much less overhead.
```
Comment (ArmaanT) [inline content/lectures/lec02.md:21]: In the second sentence, I think "build" is the wrong word since in general an image is already built from a Dockerfile. Maybe just have it reference running the code?

Comment (ArmaanT) [reply content/lectures/lec02.md:21]: Oh, just read the next paragraph. Not really sure of how to make this clearer without destroying the abstraction and talking specifically about Docker

File: content/lectures/lec02.md [hunk 3]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+## Docker
+
+Docker is the industry-standard implementation for containers. While there are other container implementations, Docker has become something like the "Kindle" or "Kleenex" of container implementations, but there [are](https://coreos.com/rkt/) [other](https://katacontainers.io) [implementations](https://firecracker-microvm.github.io) which have emerged recently and look to be giving Docker a run for its money. 
+
+The Open Container Image (OCI) format is a general specification for how to build container images originally based on the Docker image format. But other tools built against the OCI specification build compatible containers or run compatible containers, without running any code built by the Docker team. Implementations like [podman](https://containerd.io) and [containerd](https://containerd.io) are compatible with the OCI specification while improving on some ambiguities in the original Docker implementation. Kubernetes has actually [moved away from the Docker implementation](https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/) recently. We'll use "OCI container", "container", and "Docker container" interchangeably throughout these notes.
```
Comment (ArmaanT) [inline content/lectures/lec02.md:31]: OCI is Open Container Initiative, not Image. Correct wording would probably be "the OCI Image format"

File: content/lectures/lec02.md (hunk 1 above)
Comment (pawalt) [inline content/lectures/lec02.md:13]: probably explicitly say that this is the question portability answers

File: content/lectures/lec02.md (hunk 3 above)
Comment (ArmaanT) [reply content/lectures/lec02.md:31]: I'd also condense the 2nd sentence to have "build or run compatible containers"

Comment (ArmaanT) [reply content/lectures/lec02.md:31]: The Kubernetes Docker thing may cause more confusion than it's worth at the moment

File: content/lectures/lec02.md [hunk 4]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
 
+# Lecture Two: Docker
+
+## Portability
+
+Recall all of the issues that we've run into trying transfer code that we developed on our local machine to a remote host in the cloud. We want to be able to take our application code, and seamlessly run it on another machine. Along with Python dependencies, we needed to make sure that system resources, including Python and Poetry themselves, were also installed as expected. How can we avoid all of the dependency and versioning issues we ran into when we manually deployed to EC2?
+
+We were able to use Poetry and get all of the *reproducibility* benefits when we deployed to EC2, but just because a build is reproducible does *not* mean that it's **portable**.
+
+The intuition here is that we essentially want to _send a copy of our local computer_ to whoever is going to be running our code.
```
Comment (pawalt) [inline content/lectures/lec02.md:17]: I don't think this line translates to paper because it's very imprecise. The intuition, in a more precise manner, is that we want to send the other person a machine will all the dependencies pre-installed

File: content/lectures/lec02.md [hunk 5]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+The Open Container Image (OCI) format is a general specification for how to build container images originally based on the Docker image format. But other tools built against the OCI specification build compatible containers or run compatible containers, without running any code built by the Docker team. Implementations like [podman](https://containerd.io) and [containerd](https://containerd.io) are compatible with the OCI specification while improving on some ambiguities in the original Docker implementation. Kubernetes has actually [moved away from the Docker implementation](https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/) recently. We'll use "OCI container", "container", and "Docker container" interchangeably throughout these notes.
+
+## Virtual Machines vs. Containers
+
+For each virtual machine, the hypervisor fully emulate every piece of hardware that exists on the machine. Every virtual machine does this emulation individually, so if we have many virtual machines they don't share the majority of their resources. VMs each run their own kernel to talk to the emulated hardware.
```
Comment (ArmaanT) [inline content/lectures/lec02.md:35]:
```suggestion
For each virtual machine, the hypervisor fully emulates every piece of hardware that exists on the machine. Every virtual machine does this emulation individually, so if we have many virtual machines they don't share the majority of their resources. VMs each run their own kernel to talk to the emulated hardware.
```

File: content/lectures/lec02.md [hunk 6]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+Instead, containers share as many resources between multiple containers and with the host OS as possible (while still maintaining privacy and separation). This gives us a huge performance increase because emulation is always slower than actual computation. This importantly includes the host OS's kernel, which talks to the real hardware on the computer. This makes Docker containers much lighter weight than equivalent VMs.
+
+![Container Diagram](/img/lec02/container_vis.png)
+
+## Docker Images
+
+Docker images ideally contain the minimal OS needed to run your application. We only want the programs that are needed to run our application and nothing more.
+
+Docker images are built in layers. When we create a Docker image, it is always extending a base image. This is achieved with a "copy-on-write" filesystem for each image. This means that the filesystem within the image is completely immutable during the build process. Instead if modifying a file in-place, Docker will create an additional image layer with your changes on top of the base image.
```
Comment (ArmaanT) [inline content/lectures/lec02.md:47]:
```suggestion
Docker images are built in layers. When we create a Docker image, it is always extending a base image. This is achieved with a "copy-on-write" filesystem for each image. This means that the filesystem within the image is completely immutable during the build process. Instead of modifying a file in-place, Docker will create an additional image layer with your changes on top of the base image.
```

File: content/lectures/lec02.md (hunk 2 above)
Comment (pawalt) [inline content/lectures/lec02.md:21]: "send it to anyone" you probably mean send the build instructions to anyone. it's unclear, so i'd be explicit.

File: content/lectures/lec02.md [hunk 7]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+![Docker Layering Diagram](/img/lec02/container_vis.png)
+
+Copy-on-write saves us a lot of space because we only ever need one instance of a base image on a given machine. We might have ten images all running on Ubuntu, we still store only one copy of the base Ubuntu image. The ten derivative images will all create new layers on top of the shared base image.
```
Comment (ArmaanT) [inline content/lectures/lec02.md:51]: Saving space isn't a guarantee and COW removes the ability to make images smaller, so maybe we should reword this to be specific to when you have multiple images that share a base?

File: content/lectures/lec02.md [hunk 8]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+The intuition here is that we essentially want to _send a copy of our local computer_ to whoever is going to be running our code.
+
+## Containers
+
+Containers allow us to package minimal operating systems as "images" which includes everything our program needs to run. Once we've created the image, we can send it to anyone, and they will be able to build it on their machine to run our code. If you've used Virtual Machines in previous classes, you can think of container images as analogous to VM images, but with much less overhead.
+
+We can also build the image locally and send the image after it has already been built. This way the receiver can run the container without having to worry about any of the build process.
+
+Ideally, the build process itself should also be reproducible.
```
Comment (pawalt) [inline content/lectures/lec02.md:25]: isn't this what the first paragraph in this section says?

File: content/lectures/lec02.md [hunk 9]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+## Containers
+
+Containers allow us to package minimal operating systems as "images" which includes everything our program needs to run. Once we've created the image, we can send it to anyone, and they will be able to build it on their machine to run our code. If you've used Virtual Machines in previous classes, you can think of container images as analogous to VM images, but with much less overhead.
+
+We can also build the image locally and send the image after it has already been built. This way the receiver can run the container without having to worry about any of the build process.
+
+Ideally, the build process itself should also be reproducible. 
+
+## Docker
```
Comment (pawalt) [inline content/lectures/lec02.md:27]: kudos to whoever wrote this section. it's very well worded

File: content/lectures/lec02.md [hunk 10]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+A Docker container is a running instance of a Docker image. Containers still employ a copy-on-write filesystem, layered off of the image they were spawned from. This means any new, updated or removed files during the lifetime of a container are *not* persisted after a container is stopped (Side note: the picture is a bit more complex in the original Docker implementation, but clarified with most other OCI-compatible runtimes). This means if we want to save any files that are created within a container while it runs, like log files, we have to do something other than save it to the filesystem.
+
+Docker runtime runs with a single host-wide Docker [daemon](https://en.wikipedia.org/wiki/Daemon_\(computing\)) (depicted in the earlier Docker visualization) which just sits in the background, manages your containers, and processes any requests which you might make with the Docker CLI to interact with the container. Docker's daemon is considered a single point of failure, and so other runtimes manage containers without a long-running, host-wide daemon process.
```
Comment (ArmaanT) [inline content/lectures/lec02.md:63]:
```suggestion
Docker runtime runs with a single host-wide Docker [daemon](https://en.wikipedia.org/wiki/Daemon_\(computing\)) (depicted in the earlier Docker visualization) which just sits in the background, managing your containers, and processing any requests which you might make with the Docker CLI to interact with the container. Docker's daemon is considered a single point of failure so some alternative runtimes were designed to manage containers without a long-running, host-wide daemon process.
```

File: content/lectures/lec02.md [hunk 11]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+## Virtual Machines vs. Containers
+
+For each virtual machine, the hypervisor fully emulate every piece of hardware that exists on the machine. Every virtual machine does this emulation individually, so if we have many virtual machines they don't share the majority of their resources. VMs each run their own kernel to talk to the emulated hardware.
+
+![Virtual Machine Diagram](/img/lec02/vm_vis.png)
+
+Instead, containers share as many resources between multiple containers and with the host OS as possible (while still maintaining privacy and separation). This gives us a huge performance increase because emulation is always slower than actual computation. This importantly includes the host OS's kernel, which talks to the real hardware on the computer. This makes Docker containers much lighter weight than equivalent VMs.
```
Comment (pawalt) [inline content/lectures/lec02.md:39]: "share as many resources as possible" is not the right wording here. It implies that there are cases in which sharing the hardware is not possible, which is not true.

I'd move away from the "share" terminology and move towards terminology like "containers access slices of host hardware directly. the host kernel uses cgroups to securely isolate containers to independent slices of hardware.

The key idea here is that in containers, there is a single kernel that's responsible for all processes.

File: content/lectures/lec02.md [hunk 12]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+The solution here is a registry where we can publish our images so that others can download them. Common examples include:
+- Docker Hub
+- Google Container Registry (GCR)
+- Elastic Container Registry (ECR)
+- GitHub Container Registry (unfortunately also abbreviated GCR)
+
+These are essentially the same thing as NPM or PIP, but for Docker images instead of code packages.
+
+### Tags
```
Comment (ArmaanT) [inline content/lectures/lec02.md:77]: We should explicitly mention that tags are just pointers to layers

**Review by ArmaanT — CHANGES_REQUESTED** [review]

File: content/lectures/lec02.md (hunk 7 above)
Comment (pawalt) [inline content/lectures/lec02.md:51]:
```suggestion
Copy-on-write can save us a lot of space because we only ever need one instance of a base image on a given machine. We might have ten images all running on Ubuntu, we still store only one copy of the base Ubuntu image. The ten derivative images will all create new layers on top of the shared base image.
```

Comment (pawalt) [reply content/lectures/lec02.md:51]: saving space is not guaranteed

File: content/lectures/lec02.md [hunk 13]
```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
...
+
+## Container Runtime
+
+An important distinction in what we've been talking about so far is what the difference is between a **container** and an **image**. An **image** is a blueprint for an instance of the code you want to run. A **container** is running instance of a Docker image. The Docker **runtime** is what takes an image, creates a container and runs it. 
+
+A Docker container is a running instance of a Docker image. Containers still employ a copy-on-write filesystem, layered off of the image they were spawned from. This means any new, updated or removed files during the lifetime of a container are *not* persisted after a container is stopped (Side note: the picture is a bit more complex in the original Docker implementation, but clarified with most other OCI-compatible runtimes). This means if we want to save any files that are created within a container while it runs, like log files, we have to do something other than save it to the filesystem.
```
Comment (pawalt) [inline content/lectures/lec02.md:61]: clarify that the "something" is "use docker volumes"

**Review by pawalt — CHANGES_REQUESTED** [review]

File: content/lectures/lec02.md (hunk 2 above)
Comment (pawalt) [reply content/lectures/lec02.md:21]: see my comments on this part

File: content/lectures/lec02.md (hunk 12 above)
Comment (cphalen) [reply content/lectures/lec02.md:75]: I didn't include this as a change because I don't really know how this works. What's the difference between being a pointer to an image and being a pointer to a layer?
